samples/subsys/ipc/openamp/remote_echo/src/main.c

```diff
@@ -0,0 +1,299 @@
+	.get_features = virtio_get_features,
+	.set_features = virtio_set_features,
+	.notify = virtio_notify,
+};
+
+static K_SEM_DEFINE(data_sem, 0, 1);
+static K_SEM_DEFINE(data_rx_sem, 0, 1);
+static K_SEM_DEFINE(ept_sem, 0, 1);
+
+struct rpmsg_endpoint my_ept;
+struct rpmsg_endpoint *ep = &my_ept;
+
+static void platform_ipm_callback(void *context, u32_t id, volatile void *data)
+{
+	/* After receiving the first ipm message from the Linux master we assume
+	 * that the driver status is OK since the i.MX RPMsg Linux Driver
+	 * doesn't implement a set_status callback function.
+	 */
+	if (!virtio_status) {
+		virtio_status = VIRTIO_CONFIG_STATUS_DRIVER_OK;
+	}
+
+	k_sem_give(&data_sem);
+}
+
+int endpoint_cb(struct rpmsg_endpoint *ept, void *data,
+		size_t len, u32_t src, void *priv)
+{
+	ARG_UNUSED(priv);
+	ARG_UNUSED(src);
+	char payload[RPMSG_BUFFER_SIZE];
+
+	/* Send data back to master */
+	memset(payload, 0, RPMSG_BUFFER_SIZE);
+	memcpy(payload, data, len);
+
+	printk("echo message: %s\n", payload);
+	if (rpmsg_send(ept, (char *)data, len) < 0) {
+		printk("rpmsg_send failed\n");
+		goto destroy_ept;
+	}
+
+	k_sem_give(&data_rx_sem);
+
+	return RPMSG_SUCCESS;
+
+destroy_ept:
+	shutdown_req = 1;
+	k_sem_give(&data_rx_sem);
+	return RPMSG_SUCCESS;
+}
+
+
+static void rpmsg_service_unbind(struct rpmsg_endpoint *ept)
+{
+	ARG_UNUSED(ept);
+	rpmsg_destroy_ept(ep);
+}
+
+static void loopback_messages(void)
+{
+	while (k_sem_take(&data_rx_sem, K_NO_WAIT) != 0) {
+		int status = k_sem_take(&data_sem, K_FOREVER);
+
+		if (status == 0) {
+			virtqueue_notification(vq[1]);
+		}
+	}
+}
+
+void app_task(void *arg1, void *arg2, void *arg3)
+{
+	ARG_UNUSED(arg1);
+	ARG_UNUSED(arg2);
+	ARG_UNUSED(arg3);
+	int status = 0;
+	struct metal_device *device;
+	struct rpmsg_device *rdev;
+	struct metal_init_params metal_params = METAL_INIT_DEFAULTS;
+
+	shutdown_req = 0;
+	virtio_status = 0;
+
+	status = metal_init(&metal_params);
+	if (status != 0) {
+		printk("metal_init failed: %d\n", status);
+		return;
+	}
+
+	status = metal_register_generic_device(&shm_device);
+	if (status != 0) {
+		printk("Couldn't register shared memory device: %d\n", status);
+		metal_finish();
+		return;
+	}
+
+	status = metal_device_open("generic", SHM_DEVICE_NAME, &device);
+	if (status != 0) {
+		printk("metal_device_open failed: %d\n", status);
+		metal_finish();
+		return;
+	}
+
+	io = metal_device_io_region(device, 0);
+	if (io == NULL) {
+		printk("metal_device_io_region failed\n");
+		metal_finish();
+		return;
+	}
+
+	/* setup IPM */
+	ipm_handle = device_get_binding(IPM_LABEL);
+	if (ipm_handle == NULL) {
+		printk("device_get_binding %s failed\n", IPM_LABEL);
+		metal_finish();
+		return;
+	}
+
+	ipm_register_callback(ipm_handle, platform_ipm_callback, NULL);
+
+	status = ipm_set_enabled(ipm_handle, 1);
+	if (status != 0) {
+		printk("ipm_set_enabled failed: %d\n", status);
+		metal_finish();
+		return;
+	}
+
+	/* setup vdev */
+	vq[0] = virtqueue_allocate(VRING_SIZE);
```

Comment:
this function will do dynamic memory allocation. Just in case, if you want to avoid dynamic allocation, you can statically defined the virtqueue. As your demo is RPMSG_REMOTE, you can try static virtqueue vq[2].